blockmanager.go

```diff
@@ -1012,114 +1025,155 @@ func (b *blockManager) getCheckpointedCFHeaders(checkpoints []*chainhash.Hash,
 						peerAddr, err)
 				}
 
-				return false
+				return query.Progress{
+					Finished:   false,
+					Progressed: false,
+				}
 			}
 
 			// At this point, the response matches the query, and
-			// the relevant checkpoint we got earlier, so we should
-			// always return true so that the peer looking for the
-			// answer to this query can move on to the next query.
-			// We still have to check that these headers are next
-			// before we write them; otherwise, we cache them if
-			// they're too far ahead, or discard them if we don't
-			// need them.
-
-			// Find the first and last height for the blocks
-			// represented by this message.
-			startHeight := checkPointIndex*wire.CFCheckptInterval + 1
-			lastHeight := (nextCheckPointIndex + 1) * wire.CFCheckptInterval
-
-			log.Debugf("Got cfheaders from height=%v to "+
-				"height=%v, prev_hash=%v", startHeight,
-				lastHeight, r.PrevFilterHeader)
-
-			// If this is out of order but not yet written, we can
-			// verify that the checkpoints match, and then store
-			// them.
-			if startHeight > curHeight+1 {
-				log.Debugf("Got response for headers at "+
-					"height=%v, only at height=%v, stashing",
-					startHeight, curHeight)
+			// the relevant checkpoint we got earlier, so we'll
+			// deliver the verified headers on the headerChan.
+			// We'll also return a Progress indicating the query
+			// finished, that the peer looking for the answer to
+			// this query can move on to the next query.
+			select {
+			case headerChan <- r:
+			case <-b.quit:
+				return query.Progress{
+					Finished:   false,
+					Progressed: false,
+				}
 			}
 
-			// If this is out of order stuff that's already been
-			// written, we can ignore it.
-			if lastHeight <= curHeight {
-				log.Debugf("Received out of order reply "+
-					"end_height=%v, already written", lastHeight)
-				return true
+			return query.Progress{
+				Finished:   true,
```

Comment:
Don't we want this to be true once we've received all the headers?

Reply:
Or does this only apply to a single query, rather than a batch?

Reply:
Not sure how useful these `Finished`/`Progress` bools are given that the workers act upon a single message request and not a batch.

Reply:
They are useful in the single message/multiple request case, such as `getCFFilters`, `getBlocks` etc.

Reply:
A single `getcfilter` message will only ever return one response though. In the case of a batch filter fetch (multiple `getcfilter` messages) I would understand the purpose of these fields if a single worker was handling them, but that doesn't seem to be the case (the messages can be distributed among distinct workers).

Reply:
a `getCFilters` message specifies a start and stop height (https://github.com/btcsuite/btcd/blob/a41498d578a99c1619037746abd916176ea61052/wire/msggetcfilters.go#L20), and each height is returned as a single `MsgCFilter` (https://github.com/btcsuite/btcd/blob/a41498d578a99c1619037746abd916176ea61052/server.go#L819).

Reply:
Ah I see, thanks for clearing that up. In any case, `Progressed` and `Finished` still seem mutually exclusive.